Mathlib/CategoryTheory/Localization/Prod.lean

```diff
@@ -0,0 +1,132 @@
+/-
+Copyright (c) 2023 Joël Riou. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joël Riou
+-/
+import Mathlib.CategoryTheory.Localization.Equivalence
+
+/-!
+# Localization of product categories
+
+In this file, it is shown that if functors `L₁ : C₁ ⥤ D₁` and `L₂ : C₂ ⥤ D₂`
+are localization functors for morphisms properties `W₁` and `W₂`, then
+the product functor `C₁ × C₂ ⥤ D₁ × D₂` is a localization functor for
+`W₁.prod W₂ : MorphismProperty (C₁ × C₂)`, at least if both `W₁` and `W₂`
+contain identities. This main result is the instance `Functor.IsLocalization.prod`.
+
+The proof proceeds by showing first `Localization.Construction.prodIsLocalization`,
+which asserts that this holds for the localization functors `W₁.Q` and `W₂.Q` to
+the constructed localized categories: this is done by showing that the product
+functor `W₁.Q.prod W₂.Q : C₁ × C₂ ⥤ W₁.Localization × W₂.Localization` satisfies
+the strict universal property of the localization for `W₁.prod W₂`. The general
+case follows by transporting this result through equivalences of categories.
+
+-/
+
+universe v₁ v₂ v₃ v₄ v₅ u₁ u₂ u₃ u₄ u₅
+
+namespace CategoryTheory
+
+variable {C₁ : Type u₁} {C₂ : Type u₂} {D₁ : Type u₃} {D₂ : Type u₄}
+  [Category.{v₁} C₁] [Category.{v₂} C₂] [Category.{v₃} D₁] [Category.{v₄} D₂]
+  (L₁ : C₁ ⥤ D₁) {W₁ : MorphismProperty C₁} [W₁.ContainsIdentities]
+  (L₂ : C₂ ⥤ D₂) {W₂ : MorphismProperty C₂} [W₂.ContainsIdentities]
+
+namespace Localization
+
+namespace StrictUniversalPropertyFixedTarget
+
+variable {E : Type u₅} [Category.{v₅} E]
+  (F : C₁ × C₂ ⥤ E) (hF : (W₁.prod W₂).IsInvertedBy F)
+
+/-- Auxiliary definition for `prodLift`. -/
+noncomputable def prodLift₁ :
+    W₁.Localization ⥤ C₂ ⥤ E :=
+  Construction.lift (curry.obj F) (fun _ _ f₁ hf₁ => by
+    haveI : ∀ (X₂ : C₂), IsIso (((curry.obj F).map f₁).app X₂) :=
```

Comment:
Shouldn't we always just use `have` now instead of `haveI`?

Reply:
`haveI` inlines the term, which might lead to a cleaner definition. For proofs, it is certainly true that you can always just use `have`.